### PR TITLE
Updates to source tracking

### PR DESCRIPTION
**Pull request by fearthecowboy**
- if a value from the source model is assigned to a property in the target model, and explicit `use(...)` should not be necessary (regardless of via constructor, etc)

- explicit `trackTarget` should never be needed anymore

- looks like $refs are working now

- the `attic` is back

- the serialization of versionInfo is fixed.

### CODE REVIEW

File: adl/core/model/element.ts [hunk 1]
```diff
@@ -8,17 +8,27 @@ export interface Attic extends Dictionary<any> {
 
 }
 
+function clean(this: any, key: string, value: any): any {
+  return value === undefined || value === null ? value : value.valueOf();
+}
+
 export class ElementArray<T> extends Array<T> {
-  __set = new Set<string>();
+  #set = new Set<string>();
```
Comment (fearthecowboy) [inline adl/core/model/element.ts:16]: I think `ElementArray` is only temporary -- I'll probably solve unique values another way soon.

File: adl/core/test/load-openapi.ts [hunk 2]
```diff
@@ -23,7 +22,9 @@ describe('Load Single OAI3 files', () => {
   });
 
   for (const file of files) {
-
+    // if (file !== 'petstore.yaml') {
+    //  continue;
+    // }
```
Comment (nguerrera) [inline adl/core/test/load-openapi.ts:27]: Delete?

File: adl/sourcemap/proxies.ts [hunk 3]
```diff
@@ -144,15 +172,30 @@ export function using<T>(sourceValue: any, actualTargetValue: T): T | undefined
...
+  }
+  return sourceRegistry.get(instance) || targetRegistry.get(instance) || instance;
+}
+
+export function trackSource<T extends Object>(instance: T, origin: Origin, parent?: TrackedSource): T {
+  if (isSourceProxy(instance)) {
+    return instance;
+  }
+  if (isTargetProxy(instance)) {
+    debugger;
```
Comment (nguerrera) [inline adl/sourcemap/proxies.ts:192]: Intended to be temporary?

File: adl/sourcemap/proxies.ts [hunk 4]
```diff
@@ -237,12 +323,14 @@ function mkOnAdd<T>(proxy: T) {
           tracker.add(pathInTarget, o);
         }
 
-
-
         for (const { key, value } of items(<any>prop)) {
-          const raw = (<any>value);
+          if (key === 'name') {
+            debugger;
```
Comment (nguerrera) [inline adl/sourcemap/proxies.ts:328]: Intended to be temporary?

File: adl/sourcemap/proxies.ts [hunk 5]
```diff
@@ -262,6 +350,13 @@ function mkOnAdd<T>(proxy: T) {
 }
 
 export function trackTarget<T extends Object>(instance: T, pathInTarget?: Path, tracker?: Tracker): T {
+  if (isSourceProxy(instance)) {
+    debugger;
```
Comment (nguerrera) [inline adl/sourcemap/proxies.ts:354]: Again here?

File: adl/sourcemap/proxies.ts [hunk 6]
```diff
@@ -310,23 +405,29 @@ export function trackTarget<T extends Object>(instance: T, pathInTarget?: Path,
 
             case SpecialProperties.Destination:
               return targetRegistry.get(instance)!.pathInTarget;
+
+            case SpecialProperties.IsUsed:
+              debugger;
```
Comment (nguerrera) [inline adl/sourcemap/proxies.ts:410]: Here too?

File: adl/core/serialization/openapiv3/schema.ts [hunk 7]
```diff
@@ -369,16 +369,16 @@ export async function processObjectSchema(schema: v3.Schema, $: Context): Promis
...
 
-  for (const { key: propertyName, value: property } of items(schema.properties)) {
-
+  for (const { key: propertyName, value: property } of items(use(schema.properties))) {
     // process schema/reference inline
     const propSchema = await $.processPossibleReference(processSchemaReference, processSchema, property) || $.api.schemas.Any;
 
     // remove empty property
-    use(property);
+    // use(property);
```
Comment (nguerrera) [inline adl/core/serialization/openapiv3/schema.ts:377]: Delete?

Comment (nguerrera) [reply adl/core/serialization/openapiv3/schema.ts:377]: Are there more uses in this file to remove?

File: adl/core/test/scenarios/v3/single/output/ablyio.api.yaml [hunk 8]
```diff
@@ -3,58 +3,23 @@ http:
...
-        type: string
+        &a23
+        target:
+          &a1
+          type: string
+          versionInfo:
+            - added: 1.1.0
         versionInfo:
           - added: 1.1.0
-          - added: 1.1.0
```
Comment (nguerrera) [inline adl/core/test/scenarios/v3/single/output/ablyio.api.yaml:21]: Yay!

File: adl/core/test/load-openapi.ts (hunk 2 above)
Comment (fearthecowboy) [reply adl/core/test/load-openapi.ts]: fixed

File: adl/sourcemap/proxies.ts (hunk 3 above)
Comment (fearthecowboy) [reply adl/sourcemap/proxies.ts]: dammit. I thought I got all those.

Comment (fearthecowboy) [reply adl/sourcemap/proxies.ts]: fixed

File: adl/sourcemap/proxies.ts (hunk 4 above)
Comment (fearthecowboy) [reply adl/sourcemap/proxies.ts]: fixed

File: adl/sourcemap/proxies.ts (hunk 5 above)
Comment (fearthecowboy) [reply adl/sourcemap/proxies.ts]: fixed

File: adl/sourcemap/proxies.ts (hunk 6 above)
Comment (fearthecowboy) [reply adl/sourcemap/proxies.ts]: fixed

File: adl/core/serialization/openapiv3/schema.ts (hunk 7 above)
Comment (fearthecowboy) [reply adl/core/serialization/openapiv3/schema.ts]: yeah, I'll go do that now.

**Comment by fearthecowboy** [comment]
I think there is still an edge-case around the automagic use tracking, but I'll look into that later. 

At least this completely unblocks the work now.

**Pull request merged**